Disabled CI badge

jakirkham opened the pull request:
Fixes https://github.com/conda-forge/conda-smithy/issues/119

This places a disabled CI badge in place of any CI that is not operational. It also removes any link to that CI. As soon as the `skip` condition in the `meta.yaml` changes and the recipe is re-rendered, the badge will change back to the CI status and properly link to it. This basically works and is ready to go. Here is an [example]( https://github.com/conda-forge/inkscape-feedstock/blob/fde93d04d76b09ae135314752ed82885ad1c54cd/README.md ).

cc @ocefpaf @pelson @janschulz

jakirkham commented:
Sad news guys. Relative SVGs are not supported. See this issue ( https://github.com/isaacs/github/issues/316 ) for details. Note that is an unofficial GitHub issue tracker run by other users.

jakirkham commented:
So, I added a workaround for the relative SVG issue. To see it at work, take a look at this [README]( https://github.com/jakirkham/inkscape-feedstock/blob/test_disabled_ci_badge/README.md ). Basically it uses rawgit to retrieve the SVG and it requires a branch so I have set it to `master`. It is not ideal but works without having to use a crummy image. These changes have been added to this PR.

jakirkham commented:
So here are my main questions. Though others may be raised.

How do we pass state from the CI steps to the Readme in the rendering step?

1. Muck with `forge_config`/`config` to denote which CIs are building.
2. Return some sort of status/info from each CI step.
3. Check which CI files exist and update the Readme accordingly.

Here I have chosen to do 1 as it was really easy. 2 could work. 3 sounds hacky.

How do we want to handle images?

1. Go directly to [shields.io]( http://shields.io/ ).
  * Pro: Can use SVG out of the box.
  * Con: They have a long history of loading issues. We would probably make them worse.
2. Use the SVG image from the repo.
  * Pro: Hopefully a bit more stable.
  * Con: GitHub fails to render relative SVGs. Need to use [RawGit]( http://rawgit.com/ ) with a designated branch.
3. Use another image format:
  * Pro: Able to use relative reference in repo.
  * Con: It is not SVG. May look bad in different settings. Binary blob in the repo.

Here I have chosen 2. 3 could work, but is gross. Strongly discourage 1.

pelson commented on conda_smithy/templates/README.md.tmpl:
```diff
@@ -71,10 +71,22 @@ Current build status
+{%- if "circle" in ci_enabled %}
 Linux: [![Circle CI](https://circleci.com/gh/{{github.user_or_org}}/{{github.repo_name}}.svg?style=svg)](https://circleci.com/gh/{{github.user_or_org}}/{{github.repo_name}})
+{%- else %}
+Linux: ![](https://cdn.rawgit.com/{{github.user_or_org}}/{{github.repo_name}}/master/ci_support/disabled.svg)
```
We don't need this to live in each feedstock. Let's put it in the conda-smithy repo, and all feedstocks can point to it from there.

jakirkham replied:
See this [comment]( https://github.com/conda-forge/conda-smithy/pull/228#issuecomment-233098249 ). Basically relative SVGs don't work. 😒 Feel free to share your thoughts on what you think we should do about it.

jakirkham replied:
Ah sorry I misread this earlier. Yeah, we could do that. One thought though. How do we handle the fact that it may have moved location in the future? Should we point to a particular commit? Other thoughts?

jakirkham replied:
Also, I think I tried using `raw.githubusercontent.com` to reference the SVG at one point, but it renders this as text instead of an image.

pelson replied:
If it moves location in the future, we will need to rerender the feedstock (using conda-smithy). I think that is totally reasonable - so let's point at `https://cdn.rawgit.com/conda-forge/conda-smithy/master/static/disabled.svg`.

jakirkham replied:
So the one thing that I don't love about that solution is there is a lag between when the file gets moved and when a feedstock is re-rendered. We may decide to move the file, but not create a release. The net result being that all feedstocks have a temporary outage for this badge.

If we pick a commit, it won't have this problem. If we change the location and commit in conda-smithy, it will also be updated when we re-render. Though it will have no effect in-between merging and re-rendering.

jakirkham replied:
If having a sha around sounds unpleasant, maybe we could use a tag instead. That way we can use version info to fill it in.

pelson replied:
You are right, but honestly I'm not worried about it. We can always hold a copy of the SVG in the location for a period of time to give us grace on moving the feedstocks over. I don't think we need to bikeshed this one - personally I'd rather be able to move quickly on updating the SVG than worry about pointing at a specific release. Either way works for me though - you can have the final call on it.

jakirkham replied:
> You are right, but honestly I'm not worried about it.

Sorry, I'm probably being overly concerned about this. I guess I just don't want to be dealing with issues because of this badge.

> We can always hold a copy of the SVG in the location for a period of time to give us grace on moving the feedstocks over.

Yeah, that is definitely the other option and I thought about that too. I guess I wanted to free us from thinking about such silly things. There are enough problems we have to handle at any one time that I would rather the badges not be one. 😉

> I don't think we need to bikeshed this one

Sorry about that. Somehow I knew this PR was going to result in bikeshedding. Didn't realize I was going to be the one doing it though.

> personally I'd rather be able to move quickly on updating the SVG than worry about pointing at a specific release.

Yeah, I see your point. Honestly I think I'm being overly worried about the badge changing or moving. It probably is going to sit there and be the same.

> Either way works for me though - you can have the final call on it.

Ok, I have picked the commit the badge was added in. That seems to work ok FWICT. Will show in an example.

Also, did some clean-up so that the URL can be set in one place.

pelson commented on conda_smithy/configure_feedstock.py:
```diff
@@ -319,6 +322,7 @@ def main(forge_file_directory):
     recipe_dir = 'recipe'
     config = {'docker': {'image': 'condaforge/linux-anvil', 'command': 'bash'},
               'templates': {'run_docker_build': 'run_docker_build_matrix.tmpl'},
+              'ci_enabled': [],
```
Wondering whether to pivot this around and put an enabled key in the respective config items.

The template would then be:

```
{% if travis.enabled %} 
```

jakirkham replied:
Wouldn't that make it a user tunable parameter in two places though? Or should we remove it before `conda-forge.yml` is written?

jakirkham replied:
In any event, I'm ok with this suggestion. Just trying to think out what this means and how we go forward with it.

pelson replied:
Definitely don't want it to be tuneable - is ci_enabled not tuneable though?

jakirkham replied:
Good question. Not sure. This is my first time mucking with the config.

jakirkham replied:
So I have updated this to use `<CI>.enabled` format. If there was a parameter, we definitely overwrite it with these new changes. Though maybe we should be issue a warning if a user tries to set this.

jakirkham replied:
Have added code that strips `<CI>.enabled` from the `config`. If it find `<CI>.enabled` for any CI, it issues a warning that explains how to configure `skip`.

jakirkham commented:
So, updated the badge test. Benefit of point at `conda-smithy` is we can test this easily. See this [README]( https://github.com/jakirkham/inkscape-feedstock/blob/18eb8098593417167aa528d47284abcc915801bd/README.md ) for an example.

jakirkham commented:
So, updated the test run. Benefit of pointing at `conda-smithy` is we can test this easily. See this [README]( https://github.com/conda-forge/inkscape-feedstock/blob/2b6b80faac9a79880707e8909f8c2086f7a09f89/README.md ) for an example. Used the latest changes in this PR to do it.

jakirkham commented:
I think I have addressed the concerns that we have discussed, @pelson. Also have re-tested. Please let me know if there is anything else you would like me to do. If not, I think this is ready to go.

jakirkham commented:
The wording in the badge was backward. :facepalm: Have now fixed it in the badge, updated the commit, updated the example, and updated the links to the example.

jakirkham commented:
Seems to be copying the badge into feedstocks still. Will try to take a look later.

jakirkham commented:
Ok, updated the PR to ensure the badge does not get added to the feedstock. Updated my test from scratch to verify this worked and it does. This should now be ready.

jakirkham commented:
Anything else needed from me here?

jakirkham commented:
Since you seem to be rapidly merging PRs at conda-smithy @pelson, how do you feel about this one? 😉

pelson commented on conda_smithy/feedstock_content/ci_support/disabled.svg:
```diff
@@ -0,0 +1,20 @@
+<svg xmlns="http://www.w3.org/2000/svg" width="92" height="20">
```
We probably need to move this so that it isn't copied by default into a feedstock, right?

jakirkham replied:
Nope. It is not copied thanks to this [line]( https://github.com/conda-forge/conda-smithy/pull/228/files#diff-3a2c4a048580db27ad9db93e2780b511R304 ).

pelson replied:
I would probably put this somewhere else (`conda_smithy/static/disabled.svg` or something), but otherwise, 👍 .

jakirkham replied:
So, I'm not against the idea of moving it. Unfortunately moving it around does take a bit of work. Mainly because it means updating the examples. Could I convince you to save such a change for a subsequent PR?

pelson replied:
Sure can. Did you want me to do the follow-up, or are you OK with doing it?

jakirkham replied:
Either way. If you want to open an issue and assign it to me, that would work.

jakirkham replied:
Please see PR ( https://github.com/conda-forge/conda-smithy/pull/250 ) for relocation.

pelson commented:
No problem in principle. Happy to merge this if the svg doesn't find its way into a feedstock (I didn't check, but it looks like it may do currently)

jakirkham commented on conda_smithy/configure_feedstock.py:
```diff
@@ -292,7 +298,11 @@ def copytree(src, dst, ignore=(), root_dst=None):
+    copytree(
+        feedstock_content,
+        forge_dir,
+        ['README', 'ci_support/disabled.svg']
```
Was also concerned that it might be copying it to the feedstock still a couple weeks back. Adding this line ensures the badge is not copied to the feedstock. Have already tested that it works.

jakirkham commented:
Resolved merge conflicts caused by PR ( https://github.com/conda-forge/conda-smithy/pull/248 ).


Updated the Inkscape test re-rendering. Also tried a fresh re-rendering with `pyelftools` in PR ( https://github.com/conda-forge/pyelftools-feedstock/pull/7 ). Take a look at the [README]( https://github.com/jakirkham/pyelftools-feedstock/blob/428370c23755e4bd8fc47fd521c1005d85b599b2/README.md#current-build-status ). This second example was done to verify the CircleCI badge came out correctly. Also both examples were updated to verify everything still worked correctly and the disabled badge was still not being copied to the feedstocks.

jakirkham commented:
Thanks @pelson.
